editoast/src/views/v2/train_schedule/projection.rs

```diff
@@ -0,0 +1,599 @@
+        ..
+    } = data.into_inner();
+    let path_projection = PathProjection::new(&path_track_ranges);
+    let query_props = params.into_inner();
+    let train_ids = query_props.ids;
+    let infra_id = query_props.infra;
+    let conn = &mut db_pool.clone().get().await?;
+    let mut redis_conn = redis_client.get_connection().await?;
+
+    let infra = Infra::retrieve_or_fail(conn, infra_id, || TrainScheduleError::InfraNotFound {
+        infra_id,
+    })
+    .await?;
+
+    let train_schedule_batch: Vec<TrainSchedule> =
+        TrainSchedule::retrieve_batch_or_fail(conn, train_ids, |missing| {
+            TrainScheduleError::BatchTrainScheduleNotFound {
+                number: missing.len(),
+            }
+        })
+        .await?;
+    let train_map: HashMap<i64, TrainSchedule> = train_schedule_batch
+        .into_iter()
+        .map(|ts| (ts.id, ts))
+        .collect();
+
+    // 1. Retrieve cached projection
+    let mut hit_cache: HashMap<i64, CachedProjectPathTrainResult> = HashMap::new();
+    let mut miss_cache = HashMap::new();
+
+    for (train_id, train) in &train_map {
+        let pathfinding_result = pathfinding_from_train(
+            conn,
+            &mut redis_conn,
+            core_client.clone(),
+            &infra,
+            train.clone(),
+        )
+        .await?;
+
+        let track_ranges = match pathfinding_result {
+            PathfindingResult::Success {
+                track_section_ranges,
+                ..
+            } => track_section_ranges,
+            _ => continue,
+        };
+
+        let simulation_result = train_simulation(
+            db_pool.clone(),
+            redis_client.clone(),
+            core_client.clone(),
+            train,
+            &infra,
+        )
+        .await?;
+        let CompleteReportTrain {
+            report_train,
+            signal_sightings,
+            zone_updates,
+            ..
+        } = match simulation_result {
+            SimulationResult::Success { final_output, .. } => final_output,
+            _ => continue,
+        };
+        let ReportTrain {
+            times, positions, ..
+        } = report_train;
+
+        let train_details = TrainSimulationDetails {
+            positions,
+            times,
+            signal_sightings,
+            zone_updates,
+            train_path: track_ranges,
+        };
+
+        let hash = train_projection_input_hash(
+            infra.id,
+            &infra.version,
+            &train_details,
+            &path_track_ranges,
+        );
+        let projection: Option<CachedProjectPathTrainResult> = redis_conn
+            .json_get_ex(&hash, CACHE_PROJECTION_EXPIRATION)
+            .await?;
+        if let Some(cached) = projection {
+            hit_cache.insert(*train_id, cached);
+        } else {
+            miss_cache.insert(*train_id, train_details);
+        }
+    }
+    if miss_cache.is_empty() {
+        info!("Hit cache");
+    }
```

Comment:
```suggestion
    if miss_cache.is_empty() {
        info!("Hit cache for all projection parameters");
    }
```

that sounds like a steam success 😅